src/misc/degree_sequence.cpp

```diff
@@ -68,123 +70,302 @@ template<typename T> inline bool degree_less(const T &a, const T &b) {
-                VECTOR(*edges)[2 * (ec + i) + 1] = vertices[i].vertex;
-            }
+    igraph_integer_t get_prev(igraph_integer_t degree) {
+        igraph_integer_t curr = buckets[degree].prev;
+        // might need to add a check to make sure degree > 0
+        while (curr > 0 && buckets[curr].is_empty()) {
+            remove_bucket(curr);
+            curr = buckets[degree].prev;
         }
+        return curr;
+    }
 
-        ec += vd.degree;
+    igraph_integer_t get_max_bucket() {
+        // TODO: either change get_prev to take a BNode, or change
+        // head()/tail() to return integers
+        return get_prev(n_buckets - 1);
     }
 
-    return IGRAPH_SUCCESS;
+    igraph_integer_t get_min_bucket() {
+        igraph_integer_t curr = head().next;
+        while (curr < n_buckets - 1 && buckets[curr].is_empty()) {
+            remove_bucket(curr);
+            curr = head().next;
+        }
+        return curr;
+    }
 
-fail:
-    IGRAPH_ERROR("The given degree sequence cannot be realized as a simple graph.", IGRAPH_EINVAL);
-}
+    void remove_bucket(igraph_integer_t degree) { // assuming sentinel nodes
+        igraph_integer_t& prev_idx = buckets[degree].prev;
+        igraph_integer_t& next_idx = buckets[degree].next;
+        if (prev_idx != -1) buckets[prev_idx].next = next_idx;
+        if (next_idx != -1) buckets[next_idx].prev = prev_idx;
 
+        prev_idx = -1;
+        next_idx = -1;
+    }
 
-// Choose vertices in the order of their IDs.
-static igraph_error_t igraph_i_havel_hakimi_index(const igraph_vector_int_t *deg, igraph_vector_int_t *edges) {
-    igraph_integer_t n = igraph_vector_int_size(deg);
+    void insert_bucket(igraph_integer_t degree) { // assuming sentinel nodes
+        igraph_integer_t& prev_idx = buckets[degree].prev;
+        igraph_integer_t& next_idx = buckets[degree].next;
 
-    igraph_integer_t ec = 0; // number of edges added so far
+        if (prev_idx == -1 && next_idx == -1) {
+            next_idx = degree + 1;
+            prev_idx = buckets[next_idx].prev;
 
-    typedef std::list<vd_pair> vlist;
-    vlist vertices;
-    for (igraph_integer_t i = 0; i < n; ++i) {
-        vertices.push_back(vd_pair(i, VECTOR(*deg)[i]));
+            buckets[next_idx].prev = degree;
+            buckets[prev_idx].next = degree;
+        }
     }
 
-    std::vector<vlist::iterator> pointers;
-    pointers.reserve(n);
-    for (auto it = vertices.begin(); it != vertices.end(); ++it) {
-        pointers.push_back(it);
+    void insert_node(vd_pair node) {
+        insert_bucket(node.degree); // does nothing if already exists
+        buckets[node.degree].nodes.push(vd_pair{node.vertex, node.degree});
+        buckets[node.degree].count++;
     }
 
-    for (const auto &pt : pointers) {
-        vertices.sort(degree_greater<vd_pair>);
+    bool get_max_node(vd_pair& max_node) {
+        igraph_integer_t max_bucket = get_max_bucket();
+        if (max_bucket <= 0) return false;
```

Comment:
Can you write explanatory comments for when `get_max_bucket()` will return an index below 0?